go-ethereum-guides/write-basic-quiz-dapp-in-go/write-basic-quiz-dapp-in-go.md

```diff
@@ -0,0 +1,1069 @@
+contractInstance.SendAnswer(&bind.TransactOpts{
+        From: auth.From,
+        Nonce: nil,           // nil uses nonce of pending state
+        Signer: auth.Signer,
+        Value: big.NewInt(0),
+        GasPrice: nil,        // nil automatically suggests gas price
+        GasLimit: 0,          // 0 automatically estimates gas limit
+    },
+    answer,
+    )
+contractInstance.Question(&bind.CallOpts{
+    Pending: true,
+    From: auth.From,
+    Context: context.Background(),
+})
+contractInstance.CheckBoard(&bind.CallOpts{
+    Pending: true,
+    From: auth.From,
+    Context: context.Background(),
+})
+```
+
+Instead, we do it once when we create a new session:
+
+```go
+auth, _ := bind.NewTransactor(keystorefile, keystorepass)
+session := quiz.QuizSession{
+    Contract: contractInstance,
+    CallOpts: bind.CallOpts{
+        Pending: true,        // Acts on pending state if set to true
+        From: auth.From,
+        Context: context.Background(),
+    },
+    TransactOpts: bind.TransactOpts{
+        From: auth.From,
+        Nonce: nil,           // nil uses nonce of pending state
+        Signer: auth.Signer,
+        Value: big.NewInt(0),
+        GasPrice: nil,        // nil automatically suggests gas price
+        GasLimit: 0,          // 0 automatically estimates gas limit
+    },
+}
+
+session.SendAnswer(answer)
+session.Question()
+session.CheckBoard()
+```
+
+**NOTE:** `bind.NewTransactor()` returns a `bind.TransactOpts` struct with
+the `From` and `Signer` fields filled in with information from your keystore file,
+and the other fields filled in with safe defaults. You can use it as-is for
+your transactions; for example: `contractInstance.SendAnswer(auth, answer)` also
+works for our above example.
+
+Let's create a `NewSession()` function that creates a new usable session and returns it:
+
+```go
+func NewSession(ctx context.Context) (session quiz.QuizSession) {
+	loadEnv()
+	keystore, err := os.Open(myenv["KEYSTORE"])
+	if err != nil {
+		log.Printf(
+			"could not load keystore from location %s: %v\n",
+			myenv["KEYSTORE"],
+			err,
+		)
+	}
+	defer keystore.Close()
+
+	keystorepass := myenv["KEYSTOREPASS"]
+	auth, err := bind.NewTransactor(keystore, keystorepass)
+	if err != nil {
+		log.Printf("%s\n", err)
+	}
+
+	// Return session without contract instance
+	return quiz.QuizSession{
+		TransactOpts: *auth,
+		CallOpts: bind.CallOpts{
+			From:    auth.From,
+			Context: ctx,
+		},
+	}
+}
+```
+
+Here, we're:
+
+- Loading our environment variables from `.env`.
+- Reading from our keystore file.
+- Getting our keystore passphrase from the environment variable `KEYSTOREPASS`.
+- Creating a new transactor with a `bind.NewTransactor()` call.
+- Forming and returning a new `quiz.QuizSession` struct with our newly created transactor and a `CallOpts` struct with some defaults.
+
+We can then create a new session in `main()`:
+
+```go
+func main(){
+    // ...
+    session := NewSession(context.Background())
```

Comment:
@zeddee Is this at the top or the bottom on `main`? Or somewhere else?

Reply:
bottom of (what is thus far) your `func main() {}` block; I could make it clearer by showing the whole `func main()` block as it should be at this point of the guide?